src/OpenIddict.EntityFramework/Stores/OpenIddictUserStore.cs

```diff
@@ -0,0 +1,165 @@
+        /// A <see cref="Task"/> that can be used to monitor the asynchronous operation,
+        /// whose result returns the unique identifier associated with the token.
+        /// </returns>
+        public virtual async Task<string> CreateTokenAsync(TUser user, string type, CancellationToken cancellationToken) {
+            if (user == null) {
+                throw new ArgumentNullException(nameof(user));
+            }
+
+            if (string.IsNullOrEmpty(type)) {
+                throw new ArgumentException("The token type cannot be null or empty.");
+            }
+
+            // Ensure that the key type can be serialized.
+            var converter = TypeDescriptor.GetConverter(typeof(TKey));
+            if (!converter.CanConvertTo(typeof(string))) {
+                throw new InvalidOperationException($"The '{typeof(TKey).Name}' key type is not supported.");
+            }
+
+            var token = new TToken { Type = type };
+            user.Tokens.Add(token);
+
+            Context.Update(user);
+
+            await Context.SaveChangesAsync(cancellationToken);
+
+            return converter.ConvertToInvariantString(token.Id);
+        }
+
+        /// <summary>
+        /// Creates a new token associated with the given user and
+        /// attached to the tokens issued to the specified client.
+        /// </summary>
+        /// <param name="user">The user.</param>
+        /// <param name="client">The application.</param>
+        /// <param name="type">The token type.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> that can be used to abort the operation.</param>
+        /// <returns>
+        /// A <see cref="Task"/> that can be used to monitor the asynchronous operation,
+        /// whose result returns the unique identifier associated with the token.
+        /// </returns>
+        public virtual async Task<string> CreateTokenAsync(TUser user, string client, string type, CancellationToken cancellationToken) {
+            if (user == null) {
+                throw new ArgumentNullException(nameof(user));
+            }
+
+            if (string.IsNullOrEmpty(client)) {
+                throw new ArgumentException("The client identifier cannot be null or empty.");
+            }
+
+            if (string.IsNullOrEmpty(type)) {
+                throw new ArgumentException("The token type cannot be null or empty.");
+            }
+
+            // Ensure that the key type can be serialized.
+            var converter = TypeDescriptor.GetConverter(typeof(TKey));
+            if (!converter.CanConvertTo(typeof(string)) || !converter.CanConvertFrom(typeof(string))) {
+                throw new InvalidOperationException($"The '{typeof(TKey).Name}' key type is not supported.");
+            }
+
+            var key = (TKey) converter.ConvertFromInvariantString(client);
+
+            var application = await Applications.Where(entity => entity.Id.Equals(key))
```

Comment:
this can be shortened to `Applications.FirstOrDefaultAsync(entity => entity.Id.Equals(key), cancellationToken)`